### DISCUSSION TITLE
Server entrypoint to use reflect-metadata

### DISCUSSION THREAD

**Discussion by tobiasdiez** [nuxt/nuxt#15977] [Questions] [answered]
Some dependency injection frameworks require that 'import "reflect-metadata"' is added at the top of the "entry point". With nuxt 2, I've added it to the `api/index.ts` file (where api is registered as the serverMiddleware) and everything worked fine.  
But trying to use it with the new nuxt-bridge, I get the following error 
```
 WARN  [worker] tsyringe requires a reflect polyfill. Please add 'import "reflect-metadata"' to the top of your entry point.                                                                                                     18:13:20  

  at Object.<anonymous> (node_modules\tsyringe\dist\cjs\index.js:5:11)
  at Module._compile (node:internal/modules/cjs/loader:1095:14)
  at Object.Module._extensions..js (node:internal/modules/cjs/loader:1147:10)
  at Module.load (node:internal/modules/cjs/loader:975:32)
  at Function.Module._load (node:internal/modules/cjs/loader:822:12)
  at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:190:29)
  at ModuleJob.run (node:internal/modules/esm/module_job:185:25)
  at async Promise.all (index 0)
  at async ESMLoader.import (node:internal/modules/esm/loader:281:24)
  at async loadESM (node:internal/process/esm_loader:88:5)
```
Its not telling me in which file this occurs, but I guess it tries to import another file using `tsyringe` (the DI framework) before it imports the index file.

**Comment by tobiasdiez** [comment, marked as answer]
I've investigated this a bit more, and the problem seems to be less with reflect-metadata than esbuild. The latter doesn't support emitDecoratorMetadata and thus there is no metadata emitted. I've opened https://github.com/nuxt/framework/discussions/1616 to discuss how one could configure esbuild to fix this.

**Reply by MiraiSubject** [reply]
The solution described in nuxt/nuxt#15905 does not work anymore on `nuxt3@3.0.0-27500677.83d27bd` 

```typescript
import { defineNuxtConfig } from 'nuxt3'

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
    nitro: {
        hooks: {
            'nitro:rollup:before'(ctx) {
                ctx.options.moduleSideEffects.push('reflect-metadata')
            }
        }
    }
})
```
After applying this to my `nuxt.config.ts` I still get the same error. In my case tsyringe gets imported inside `server/middleware/express.ts`. Adding `import 'reflect-metadata` at the top of `express.ts` does not solve it either. 

My `tsconfig.json`: 
```json
{
  // https://v3.nuxtjs.org/concepts/typescript
  "extends": "./.nuxt/tsconfig.json",
  "compilerOptions": {
    "esModuleInterop": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
  }
}
```

**Reply by bigBron** [reply]
> nuxt3 sever side import 'reflect-metadata' using this configuration
```typescript
// nuxt.config.ts
import {Nitro} from "nitropack";

export default defineNuxtConfig({
    hooks: {
        'nitro:build:before': (nitro: Nitro) => {
            nitro.options.moduleSideEffects.push('reflect-metadata')
        }
    }
})
```

**Reply by jpujol880807** [reply]
Any solutions?, I'm still getting `[nitro] [uncaughtException] TypeError: Reflect.hasOwnMetadata is not a function` using last comment

**Reply by mnaviddoost** [reply]
Thank you  very  very  very  much

**Comment by pipinet** [comment]
```javascript
nitro: {
    hooks: {
      'rollup:before': (nitro) =>{
        nitro.options.moduleSideEffects.push('reflect-metadata')
      }
    },
    rollupConfig:{
      output: {
        banner: 'import "reflect-metadata";',
      }
    }
  },
```
